linux/dm-sworndisk/src/lib.rs

```diff
@@ -3,36 +3,351 @@
+    sync::atomic::{AtomicBool, Ordering},
+};
+use crossbeam_queue::SegQueue;
+use kernel::{c_str, new_condvar, prelude::*, sync::CondVar};
+use sworndisk::{Arc, BlockId, BlockSet, BufMut, BufRef, Errno, Error, Mutex};
+
+use crate::bio::{Bio, BioOp, BioVec};
+use crate::block_device::{HostBlockDevice, BLOCK_SECTORS, BLOCK_SIZE};
+use crate::device_mapper::{Args, DmTargetOps, MapState, Target, TargetType};
 
 module! {
-    type: Dummy,
+    type: TargetManager,
     name: "dm_sworndisk",
     author: "Rust for Linux Contributors",
     description: "Rust dm_sworndisk module",
     license: "GPL",
 }
 
-struct Dummy;
+/// A struct used to manage a `TargetType`.
+///
+/// Register the type to kernel when `init` the module, and `unregister` it
+/// when `drop` (automatically).
+struct TargetManager(Pin<Box<TargetType>>);
 
-impl kernel::Module for Dummy {
+impl kernel::Module for TargetManager {
     fn init(_module: &'static ThisModule) -> Result<Self> {
         pr_info!("Rust dm_sworndisk module (init)\n");
 
-        test_rwlock();
-        test_weak();
-        test_thread();
+        // TODO: add a switch to enable those tests.
+        // test_rwlock();
+        // test_weak();
+        // test_thread();
 
-        Ok(Dummy)
+        let dm_sworndisk = Box::pin_init(TargetType::register::<DmSwornDisk>(
+            c_str!("sworndisk"),
+            [0, 0, 1],
+            0,
+        ))?;
+
+        Ok(TargetManager(dm_sworndisk))
     }
 }
 
-impl Drop for Dummy {
+impl Drop for TargetManager {
     fn drop(&mut self) {
         pr_info!("Rust dm_sworndisk module (exit)\n");
     }
 }
 
+/// A request queue, dispatching bios from device mapper to `RawDisk`.
+struct ReqQueue {
+    bios: Mutex<SegQueue<Bio>>,
+    // TODO: replace `RawDisk` with `SwornDisk`.
+    disk: RawDisk,
+    should_stop: AtomicBool,
+    new_bio: Pin<Box<CondVar>>,
+}
+
+impl ReqQueue {
+    /// Constructs a `ReqQueue`.
+    pub fn new(disk: RawDisk) -> Self {
+        Self {
+            bios: Mutex::new(SegQueue::new()),
+            disk,
+            should_stop: AtomicBool::new(false),
+            new_bio: Box::pin_init(new_condvar!()).unwrap(),
+        }
+    }
+
+    /// Returns true if the device mapper is going to exit.
+    fn should_stop(&self) -> bool {
+        self.should_stop.load(Ordering::Acquire)
+    }
+
+    /// Set the `should_stop` flag.
+    ///
+    /// If the device mapper target is going to exit, user should call this
+    /// method, in order to tell the `ReqQueue` handler thread to clear all
+    /// the pending bios.
+    pub fn set_stopped(&self) {
+        self.should_stop.store(true, Ordering::Release);
+        self.new_bio.notify_all();
+    }
+
+    /// Enqueues a `Bio`.
+    pub fn enqueue(&self, bio: Bio) {
+        self.bios.lock().push(bio);
+        self.new_bio.notify_one();
+    }
+
+    /// Spawns a `JoinHandle` to deal with bio requests in the queue.
+    pub fn spawn_req_handler(queue: Arc<Self>) -> JoinHandle<()> {
```

Comment:
I suggest using the term `worker` rather than `handler`. Because when we talk about a worker, we usually mean a worker thread, which is exactly the case here. And to be consistent, you do not say the worker _handles_ requests, you say the worker _processes_ requests. Rename the `handle` method to `process` or `process_req` or `process_bio`.